src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -265,15 +277,56 @@ public NodeStateTable<T> ToImmutableAndFree()
                     return NodeStateTable<T>.Empty;
                 }
 
-                var hasNonCached = _states.Any(static s => s.Any(static i => i.Item2 != EntryState.Cached));
+                var hasNonCached = _states.Any(static s => !s.IsCached);
                 return new NodeStateTable<T>(_states.ToImmutableAndFree(), isCompacted: !hasNonCached);
-
             }
 
             internal ImmutableArray<T> GetLastEntries()
             {
-                return _states[_states.Count - 1].SelectAsArray(t => t.Item1);
+                var nodeStateTableEntry = _states[_states.Count - 1];
+                return nodeStateTableEntry.Item is object ? ImmutableArray.Create<T>(nodeStateTableEntry.Item) : nodeStateTableEntry.Items;
+            }
+        }
+
+        private readonly struct TableEntry
+        {
+            internal readonly ImmutableArray<T> Items;
+
+            internal readonly T? Item;
+
+            internal readonly EntryState? State;
+
+            internal readonly ImmutableArray<EntryState> States;
```

Comment:
I'm curious if it might be simpler to store `(T, EntryState)` and `ImmutableArray<(T, EntryState)>` rather than keeping `T` and `EntryState` separate in each case.